server/fleet/software.go

```diff
@@ -170,3 +185,66 @@ func (uhsdbr *UpdateHostSoftwareDBResult) CurrInstalled() []Software {
 
 	return r
 }
+
+// ParseSoftwareLastOpenedAtRowValue attempts to parse the last_opened_at
+// software column value. If the value is empty or if the parsed time is
+// 0 it returns (time.Time{}, nil).
+func ParseSoftwareLastOpenedAtRowValue(value string) (time.Time, error) {
+	// We don't fail if only the last_opened_at cannot be parsed.
+	if value == "" {
+		return time.Time{}, nil
+	}
+	lastOpenedEpoch, err := strconv.ParseFloat(value, 64)
+	if err != nil {
+		return time.Time{}, err
+	}
+	if lastOpenedEpoch == 0 {
+		return time.Time{}, nil
+	}
+	return time.Unix(int64(lastOpenedEpoch), 0).UTC(), nil
+}
+
+// SoftwareFromOsqueryRow creates a fleet.Software from the values reported by osquery.
+// Arguments name and source must be defined, all other fields are optional.
+// This method doesn't fail if lastOpenedAt is empty or cannot be parsed.
+//
+// All fields are trimmed to fit on Fleet's database.
+// The vendor field is currently trimmed by removing the extra characters and adding `...` at the end.
+func SoftwareFromOsqueryRow(name, version, source, vendor, installedPath, release, arch, bundleIdentifier, lastOpenedAt string) (*Software, error) {
+	if name == "" {
+		return nil, errors.New("host reported software with empty name")
+	}
+	if source == "" {
+		return nil, errors.New("host reported software with empty source")
+	}
+
+	// We don't fail if only the last_opened_at cannot be parsed.
+	lastOpenedAtTime, _ := ParseSoftwareLastOpenedAtRowValue(lastOpenedAt)
+
+	// Check whether the vendor is longer than the max allowed width and if so, truncate it.
+	if utf8.RuneCountInString(vendor) >= SoftwareVendorMaxLength {
+		vendor = fmt.Sprintf(SoftwareVendorMaxLengthFmt, vendor)
+	}
+
+	truncateString := func(str string, length int) string {
```

Comment:
Note that this truncates to `length` **bytes**, not runes. This could cause issues and maybe end up in database error inserting invalid utf8 if a rune is truncated in the middle. I think it would be best to account for this, converting the string to `[]rune` and truncating that rune slice, and return it back to string.

Reply:
Great catch! Fixed in [PR fixes](https://github.com/fleetdm/fleet/pull/13741/commits/7c381c1d57d1a2a0e2f8e984eebb59b32f5dbcb0).